chia/data_layer/data_layer.py

```diff
@@ -267,37 +270,51 @@ async def fetch_and_validate(self, subscription: Subscription) -> None:
 
         root = await self.data_store.get_tree_root(tree_id=tree_id)
         # Wallet root hash must match to our data store root hash.
-        if root.node_hash is not None and root.node_hash == to_check[0].root:
+        if to_check[0].root == (root.node_hash if root.node_hash is not None else self.none_bytes):
             self.log.info(
-                f"Validated chain hash {root.node_hash} in downloaded datastore. "
+                f"Validated chain hash {to_check[0].root} in downloaded datastore. "
                 f"Wallet generation: {to_check[0].generation}"
             )
         else:
-            await self.data_store.rollback_to_generation(tree_id, (0 if old_root is None else old_root.generation))
-            raise RuntimeError("Can't find data on chain in our datastore.")
+            await self.data_store.set_validated_wallet_generation(tree_id, 0)
+            await self.data_store.rollback_to_generation(tree_id, 0)
+            raise RuntimeError("Could not validate on-chain data. Downloading from scratch as a fallback.")
+        last_checked_hash = to_check[0].root
         to_check.pop(0)
         min_generation = (0 if old_root is None else old_root.generation) + 1
         max_generation = root.generation
 
         # Light validation: check the new set of operations against the new set of wallet records.
         # If this matches, we know all data will match, as we've previously checked that data matches
         # for `min_generation` data store root and `wallet_current_generation` wallet record.
-        is_valid: bool = await self._validate_batch(tree_id, to_check, min_generation, max_generation)
+        try:
+            is_valid: bool = await self._validate_batch(
+                tree_id, to_check, last_checked_hash, min_generation, max_generation
+            )
+        except Exception as e:
```

Comment:
This consumes cancellation on Python 3.7.

Reply:
Is adding this a reasonable thing to do?
```
except asyncio.CancelledError:
    raise

Reply:
That's the basic gist, yeah.  Other than just catching exceptions you know what to do with.